crates/red_knot_python_semantic/src/types/infer.rs

```diff
@@ -2212,6 +2212,10 @@ impl<'db> TypeInferenceBuilder<'db> {
         match (op, self.infer_expression(operand)) {
             (UnaryOp::USub, Type::IntLiteral(value)) => Type::IntLiteral(-value),
             (UnaryOp::Not, Type::BooleanLiteral(value)) => Type::BooleanLiteral(!value),
+            (UnaryOp::Not, Type::IntLiteral(value)) => match value {
+                0 => Type::BooleanLiteral(true),
+                _ => Type::BooleanLiteral(false),
+            },
```

Comment:
you can simplify this slightly:

```suggestion
            (UnaryOp::Not, Type::IntLiteral(value)) => Type::BooleanLiteral(value == 0),
```